### PR TITLE
[Driver][SYCL][NewOffloadModel] Hook up -fsycl-device-only behaviors

### PR DESCRIPTION
**Pull request by mdtoguchi**
When using -fsycl-device-only with the new offloading model, update the behaviors to properly restrict the compilation flow to only produce the device binary.  This more tightly integrates with the existing --offload-device-only option that is commonly used.

### CODE REVIEW

File: clang/test/Driver/sycl-offload-new-driver.c [hunk 1]
```diff
@@ -52,3 +52,16 @@
...
+// RUN: %clangxx --target=x86_64-unknown-linux-gnu -fsycl --offload-new-driver \
+// RUN:          -fsycl-device-only -ccc-print-phases %s 2>&1 \
+// RUN   | FileCheck -check-prefix DEVICE_ONLY %s
+// RUN: %clangxx --target=x86_64-unknown-linux-gnu -fsycl --offload-new-driver \
+// RUN:          --offload-device-only -ccc-print-phases %s 2>&1 \
+// RUN:  | FileCheck -check-prefix DEVICE_ONLY %s
+// DEVICE_ONLY: 0: input, "{{.*}}", c++, (device-sycl)
+// DEVICE_ONLY: 1: preprocessor, {0}, c++-cpp-output, (device-sycl)
+// DEVICE_ONLY: 2: compiler, {1}, ir, (device-sycl)
+// DEVICE_ONLY: 3: backend, {2}, ir, (device-sycl)
```
Comment (asudarsa) [inline clang/test/Driver/sycl-offload-new-driver.c:66]: I do not see this phase in the old offloading model. What does this phase do? Thanks

Comment (mdtoguchi) [reply clang/test/Driver/sycl-offload-new-driver.c:66]: I kind of went back and forth on this one, as the new offloading model area implies both of these steps (compiler, which creates the IR, then backend which does processing after that - typically generating assembly).  The assembly generation step is redundant here, as it is combined with the compiler step so we don't actually see both of these phases unless it is purposefully split out.  I think I'll look at this a bit more and just do the compiler step, as it is a cleaner aspect of what we are actually doing.

Comment (mdtoguchi) [reply clang/test/Driver/sycl-offload-new-driver.c:66]: After additional investigation, the additional `backend` step works best with the new offload model path - especially when dealing with being able to emit textual IR with `-fsycl-device-only`.  I will be leaving the change as-is.

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -7882,9 +7882,9 @@ Action *Driver::BuildOffloadingActions(Compilation &C,
         break;
       }
 
-      // Backend/Assemble actions are not used for the SYCL device side
-      if (Kind == Action::OFK_SYCL &&
-          (Phase == phases::Backend || Phase == phases::Assemble))
+      // Assemble actions are not used for the SYCL device side.  Both compile
```
Comment (asudarsa) [inline clang/lib/Driver/Driver.cpp:7885]: Is this  going to deviate from the old-offloading-model flow where we do not seem to be calling Backend? If yes, is that agreeable? Thanks

Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:7885]: Yes, this is agreeable.  The functionality is the same, we just introduce the additional phase to better match with the new model expectations.

**Review by hchilama — APPROVED** [review]

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks

**Comment by asudarsa** [comment]
Hi @mdtoguchi 

Can you please look at the cuda test fails? 

Thanks

**Comment by asudarsa** [comment]
> Hi @mdtoguchi
> 
> Can you please look at the cuda test fails?
> 
> Thanks

Cuda fails are unrelated to this PR and a fix has been submitted for review here: https://github.com/intel/llvm/pull/13692

**Review by hdelan — APPROVED** [review]
Thanks for the ping. LGTM

**Comment by mdtoguchi** [comment]
@intel/llvm-gatekeepers, this looks ready for merge - thanks!

**Pull request merged**